unity-renderer/Assets/DCLServices/EmotesCatalog/EmotesCatalogService.cs

```diff
@@ -0,0 +1,238 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Cysharp.Threading.Tasks;
+using DCL.Helpers;
+
+public class EmotesCatalogService : IEmotesCatalogService
+{
+    private readonly IEmotesCatalogBridge bridge;
+
+    internal readonly Dictionary<string, WearableItem> emotes = new Dictionary<string, WearableItem>();
+    internal readonly Dictionary<string, HashSet<Promise<WearableItem>>> promises = new Dictionary<string, HashSet<Promise<WearableItem>>>();
+    internal readonly Dictionary<string, int> emotesOnUse = new Dictionary<string, int>();
+    internal readonly Dictionary<string, HashSet<Promise<WearableItem[]>>> ownedEmotesPromises = new Dictionary<string, HashSet<Promise<WearableItem[]>>>();
+
+    public EmotesCatalogService(IEmotesCatalogBridge bridge, WearableItem[] embeddedEmotes)
+    {
+        this.bridge = bridge;
+        EmbedEmotes(embeddedEmotes);
+    }
+
+    public void Initialize()
+    {
+        bridge.OnEmotesReceived += OnEmotesReceived;
+        bridge.OnOwnedEmotesReceived += OnOwnedEmotesReceived;
+    }
+
+    private void OnEmotesReceived(WearableItem[] receivedEmotes)
+    {
+        for (var i = 0; i < receivedEmotes.Length; i++)
+        {
+            var emote = receivedEmotes[i];
+            //If we dont have promises waiting, this emote is not needed and can be ignored
+            if (!promises.TryGetValue(emote.id, out var emotePromises) || emotePromises.Count == 0)
+            {
+                emotesOnUse.Remove(emote.id);
```

Comment:
It's because the emotes on use are updated when receiving emotes and therefore we are using the promises to know how many times it was requested.

In any case your comment made me realize a small edge case where we could have a system forgetting an emote before it was received and avoiding cleaning it up properly.

I've changed the code, now it should be WAY more cleaner and understandable. Thanks!